app/models/manageiq/providers/hawkular/middleware_manager.rb

```diff
@@ -143,6 +143,30 @@ def redeploy_middleware_deployment(ems_ref)
       run_generic_operation(:Redeploy, ems_ref)
     end
 
+    def add_middleware_deployment(ems_ref, hash)
+      with_provider_connection do |connection|
+        deployment_data = {
+          :enabled               => hash["enabled"],
+          :destination_file_name => hash["runtime_name"] || hash["file"].original_filename,
+          :binary_content        => hash["file"].read,
+          :resource_path         => ems_ref.to_s
+        }
+
+        actual_data = {}
+        connection.operations(true).add_deployment(deployment_data) do |on|
+          on.success do |data|
+            _log.debug "Success on websocket-operation #{data}"
+            actual_data[:data] = data
+          end
+          on.failure do |error|
+            actual_data[:data]  = {}
+            actual_data[:error] = error
+            _log.error 'error callback was called, reason: ' + error.to_s
```

Comment:
I'd say so, relying that the error returned from the ruby-client/ws operation should specify it. This is pretty much how it's done as well for generic operations (restart, reload server, redeploy, undeploy deployment).